### DISCUSSION TITLE
Best practice to synchronize video with other data modality

### DISCUSSION THREAD

**Discussion by KangByungwoo** [bonsai-rx/bonsai#1491] [Q&A]
Hi,

Currently, to synchronize video with other data modality, I am sending a 1Hz square wave input to the camera (Flea3 FL3-U3-13Y3M)'s GPIO Pin state and to a DAQ board where I record other signals for other data modality. The camera's GPIO Pin state is recorded in the csv file generated by CSVWriter, so by identifying the first frame in which its GPIO Pin state is changed from the baseline, I can synchronize video frames with the rest of data modality. But, the accuracy of this method is limited by the inverse of the video frame rate. Is there a better way to achieve the synchronization? If there's a way to trigger video acquisition by an external input that is recorded also by a DAQ board, that could be a solution. But, then, there could be an (unknown) delay between trigger and the first frame, and also if the first frame drops for some reason, it would lead to even more uncertainty...

My workflow is shown below.
![image](https://github.com/bonsai-rx/bonsai/assets/26239966/7df84e4b-61da-4ffa-8132-3c40210ac0ea)

**Comment by bruno-f-cruz** [comment]
Depends on the model of your camera. Computer vision cameras usually have 3 ways of achieving this:

1. Recording an input TTL on each frame (you seem to be using this one)
2. Output the strobe signal. This is a signal that goes up and down on each frame. You can control the time of the first pulse by delaying the subscription to the camera in Bonsai.
3. Set the camera to trigger mode and acquire a frame on each incoming pulse. This will probably afford you the greatest precision if what you are looking for is to synchronize multiple cameras. You can simply connect all cameras to the same trigger source.

**Reply by KangByungwoo** [reply]
Thank you for your answers! Though is there a way to deal with drop frames in Method 2 and 3? Previously, I recorded the strobe signal from the camera on a DAQ board, but the number of frames inferred from it didn't match the actual number of frames, meaning that there were drop frames. In Method 3, I can imagine that we can run into similar problems.  

Also, in Method 2, how would you "delaying the subscription to the camera in Bonsai"?

How would you do Method 3? Is there any step-by-step guide for it? Does this method involve Bonsai or just FlyCapture2?

**Reply by bruno-f-cruz** [reply]
Once again depends on your camera. In FLIR cameras, AFAIK, each frame metadada has a frame counter that can be used to infer dropped frames in both Method 2 and 3. So if you indeed dropped a frame, you can know to which Strobe / Trigger it would have corresponded to.

You can delay the subscription in several different ways. The simplest it probably to use the "SubscribeWhen" or "DelaySubscription". 

You need to configure your camera to use trigger mode (have you checked the manual?). Bonsai will connect to the camera, similarly to "free-running" mode however, frame events will only be produced once a trigger is provided to the camera.
